src/coreclr/jit/lowerxarch.cpp

```diff
@@ -4318,8 +4318,7 @@ void Lowering::ContainCheckStoreIndir(GenTreeStoreInd* node)
     // an int-size or larger store of zero to memory, because we can generate smaller code
     // by zeroing a register and then storing it.
     GenTree* src = node->Data();
-    if (IsContainableImmed(node, src) &&
-        (!src->IsIntegralConst(0) || varTypeIsSmall(node) || node->Addr()->OperIs(GT_CLS_VAR_ADDR)))
+    if (IsContainableImmed(node, src) && (!src->IsIntegralConst(0) || varTypeIsSmall(node)))
```

Comment:
Are you sure it doesn't regress a case when we e.g. set a constant value to a static field via reloc?

Reply:
Hmm, not sure exactly what case do you have in mind.

The static will be addressed with a reloc, that does not change. The only thing that changes is how we get the zero.
```diff
-       mov      dword ptr [classVar[0xd1ffab1e]], 0
+       xor      eax, eax
+       mov      dword ptr [classVar[0xd1ffab1e]], eax
```

Reply:
Ah, I see, the case where native compilers prefer always doing larger (in terms of size) `mov [mem], 0` ;-)

Reply:
What does a microbenchmark show between those two forms?

Reply:
> What does a microbenchmark show between those two forms?

I suspect `mov m, 0` is faster in a general case but if we enable "Contained" for zeros for all cases here we'll get a massive size increase (positive jit diff).
```asm
xor r, r
mov m, r
```
occupies a register and I suspect it takes more cycles to execute than `mov m, 0`.

There is even a smaller version:
```csharp
and m, 0
```
but it modifes flags => it's slower. But we might consider it for `JIT_FLAG_SIZE_OPT` mode

Reply:
I wrote a naive BDN benchmark that exercises the difference between the two forms. 

<details>
<summary>The naive benchmark</summary>

```cs
public class Bench
{
    private static int _intStatic1;
    private static int _intStatic2;
    private static int _intStatic3;
    private static int _intStatic4;
    private static int _intStatic5;
    private static int _intStatic6;
    private static int _intStatic7;
    private static int _intStatic8;
    private static int _intStatic9;
    private static int _intStatic10;
    private static int _intStatic11;
    private static int _intStatic12;
    private static int _intStatic13;
    private static int _intStatic14;
    private static int _intStatic15;
    private static int _intStatic16;


    public const int RepeatCount = 1000

    [Benchmark(Baseline = true)]
    public void Contained()
    {
        _intStatic1 = 0;
        _intStatic2 = 0;
        _intStatic3 = 0;
        _intStatic4 = 0;
        _intStatic5 = 0;
        _intStatic6 = 0;
        _intStatic7 = 0;
        _intStatic8 = 0;
        _intStatic9 = 0;
        _intStatic10 = 0;
        _intStatic11 = 0;
        _intStatic12 = 0;
        _intStatic13 = 0;
        _intStatic14 = 0;
        _intStatic15 = 0;
        _intStatic16 = 0;
    }

    [Benchmark]
    public void Uncontained()
    {
        ref var a = ref _intStatic1;
        a = 0;
        ref var b = ref _intStatic2;
        b = 0;
        ref var c = ref _intStatic3;
        c = 0;
        ref var d = ref _intStatic4;
        d = 0;
        ref var e = ref _intStatic5;
        e = 0;
        ref var f = ref _intStatic6;
        f = 0;
        ref var g = ref _intStatic7;
        g = 0;
        ref var h = ref _intStatic8;
        h = 0;

        ref var r9 = ref _intStatic9;
        r9 = 0;
        ref var r10 = ref _intStatic10;
        r10 = 0;
        ref var r11 = ref _intStatic11;
        r11 = 0;
        ref var r12 = ref _intStatic12;
        r12 = 0;
        ref var r13 = ref _intStatic13;
        r13 = 0;
        ref var r14 = ref _intStatic14;
        r14 = 0;
        ref var r15 = ref _intStatic15;
        r15 = 0;
        ref var r16 = ref _intStatic16;
        r16 = 0;
    }
}
```

(I modified the Jit to not reuse the zero for the second case - it would ordinarily do so, this is in fact the source of some of the bigger size diffs for this change)
</details>

The result of such a benchmark (most likely bottlenecked on the memory throughput) is of course of questionable value, but here it is:
```scala
|      Method |     Mean |     Error |    StdDev | Ratio | RatioSD |
|------------ |---------:|----------:|----------:|------:|--------:|
|   Contained | 2.823 ns | 0.0501 ns | 0.0469 ns |  1.00 |    0.00 |
| Uncontained | 2.892 ns | 0.0933 ns | 0.0917 ns |  1.02 |    0.03 |
```

Reply:
> I suspect mov m, 0 is faster in a general case but if we enable "Contained" for zeros for all cases here we'll get a massive size increase (positive jit diff).

That would be my guess as well, but if that is the case then the question is if we want to take a regression simply for consistency's sake? We have other cases where we prefer larger code because it's faster (e.g. all VSD call sequences are 3 bytes larger than they need to be because `mov rax, immed` is faster than `lea rax, [rip+disp]` on Intel CPUs).

Reply:
I don't have a strong opinion here but I support @SingleAccretion's intention to get rid of `CLS_VAR_ADDR`

Reply:
> e.g. all VSD call sequences are 3 bytes larger than they need to be because mov rax, immed

btw, on arm64 we have a memory load instead of it 🙁 https://github.com/dotnet/runtime/blob/main/src/coreclr/vm/arm64/virtualcallstubcpu.hpp#L112-L113

Reply:
If we believe the current containment behavior is a pessimization, I think I can reverse this change to go in the other direction and contain zeroes for all constant addresses, which would also solve the issue of diffs for `CLS_VAR` deletion.

But it will have a (potentially large) negative size impact (and maybe performance as well in cases where we can reuse the zeroes), and I don't think I have what it takes to be able to prove it is beneficial.

Finally, we could accept that we'll need a quirk to retain the previous behavior for `CLS_VAR` (something like `addr->IsIconHandle(GTF_ICON_STATIC_HDL) && !node->TypeIs(TYP_REF)`. I am not a big fan of quirks like that, personally.

Reply:
I think given the benchmark results this is completely fine.